openmc/model/surface_composite.py

```diff
@@ -50,6 +52,155 @@ def __pos__(self):
     def __neg__(self):
         """Return the negative half-space of the composite surface."""
 
+class IsogonalOctagon(CompositeSurface):
+    """Infinite isogonal octagon composite surface
+
+    An isogonal octagon is composed of eight planar surfaces. The prism is
+    parallel to the x, y, or z axis. The remaining two axes (y and z, x and z,
+    or x and y) serve as a basis for constructing the surfaces. Two surfaces
+    are parallel to the first basis axis, two surfaces are parallel
+    to the second basis axis, and the remaining four surfaces intersect both
+    basis axes at 45 degree angles.
+
+    This class acts as a proper surface, meaning that unary `+` and `-`
+    operators applied to it will produce a half-space. The negative side is
+    defined to be the region inside of the octogonal prism.
+
+    Parameters
+    ----------
+    center : iterable of float
+        Coordinate for the central axis of the octagon in the
+        (y, z), (x, z), or (x, y) basis.
+    r1 : float
+        Half-width of octagon across its basis axis-parallel sides in units
+        of cm. Assumed to be less than :math:`r_2\sqrt{2}`.
+    r2 : float
+        Half-width of octagon across its basis axis intersecting sides in
+        units of cm. Assumed to be less than :math:`r_1\sqrt{2}`.
+    axis : {'x', 'y', 'z'}
+        Central axis of octagon. Defaults to 'z'
+    **kwargs
+        Keyword arguments passed to underlying plane classes
+
+    Attributes
+    ----------
+    top : openmc.ZPlane or openmc.YPlane
+        Top planar surface of octagon
+    bottom : openmc.ZPlane or openmc.YPlane
+        Bottom planar surface of octagon
+    right: openmc.YPlane or openmc.XPlane
+        Right planar surface of octagon
+    left: openmc.YPlane or openmc.XPlane
+        Left planar surface of octagon
+    upper_right : openmc.Plane
+        Upper right planar surface of octagon
+    lower_right : openmc.Plane
+        Lower right planar surface of octagon
+    lower_left : openmc.Plane
+        Lower left planar surface of octagon
+    upper_left : openmc.Plane
+        Upper left planar surface of octagon
+
+    """
+
+    _surface_names = ('top', 'bottom',
+                      'upper_right', 'lower_left',
+                      'right', 'left',
+                      'lower_right', 'upper_left')
+
+    def __init__(self, center, r1, r2, axis='z', **kwargs):
+        self._axis = axis
+        c1, c2 = center
+
+        # Coords for axis-perpendicular planes
+        ctop = c1 + r1
+        cbottom = c1 - r1
+
+        cright = c2 + r1
+        cleft = c2 - r1
+
+        # Side lengths
+        if r2 > r1 * sqrt(2):
+            warnings.warn(f'r2 is greater than sqrt(2) * r1. Octagon may' + \
+                          ' be erroneous.')
+        if r1 > r2 * sqrt(2):
+            warnings.warn(f'r1 is greater than sqrt(2) * r2. Octagon may' + \
+                          ' be erroneous.')
```

Comment:
Is there any situation where we would actually want to allow this? If not we should just make these exceptions rather than warnings.